### ISSUE TITLE
Consider transport agnostic negotation mechanism

### ISSUE THREAD

**Issue by estan** [wamp-proto/wamp-proto#233]
I originally sent this to the mailing list, but @goeddea suggested I add it here as an issue to discuss.

```
This is not so much a question, as an observation:

For WAMP over WebSocket, WAMP uses WebSockets subprotocol negotiation
mechanism, not only to negotiate the fact that it is WAMP, but also
the variant of WAMP (serialization, batch mode).

For WAMP over RawSocket, WAMP has a more rudimentary negotiation
mechanism defined, which is kept in the spec.

I think from an implementers perspective, it's a little unfortunate
that the spec does not define its own transport protocol agnostic
negotiation mechanism for these specifics (serialization, batch mode).
If there was one single way to negotiate these things, regardless of
transport, then an implementor could write that piece of handshaking
logic once, and it would work for any current and future transports.
It would also have the added benefit of not having to have anything
more than "wamp" registered at IANA.

Moreover, subprotocol negotiation is a non-normative part of the
WebSocket spec, and some WebSocket implementations have chosen or
neglected to implement it.

Has the idea of having one single transport agnostic negotiation
mechanism for serialization / batch mode been discussed? (It would be
an elaboration I guess of the current one for RawSocket).
```

**Comment by oberstet** [comment]
> subprotocol negotiation is a non-normative part of the WebSocket spec

No, it's mandatory. https://tools.ietf.org/html/rfc6455#section-4.2.2

> some WebSocket implementations have chosen or neglected to implement it

Then those implementations are non-conforming

> one single transport agnostic negotiation mechanism for serialization / batch mode been discussed

No. And I a) don't see how that would work (because how would a WAMP opening handshake serialize messages if the serializer isn't known or negotiated beforehand) and b) I think it would break the abstraction of WAMP from transports

In my experience, the latter has proven to be extremely valuable/convenient.

**Comment by oberstet** [comment]
"non-normative" sections in an RFC are just simplified, colloquial description, but there will be a corresponding "normative" section. Both termini are orthogonal to MUST, MAY or SHOULD wording ..

**Comment by meejah** [comment]
Also, the RawSocket stuff isn't really "negotiation", it's just asking for something and it fails if the server doesn't support it; the only "negotiation" would come if the client keeps re-trying with less- and less- preferred options. In the HTTP case, this would mean multiple HTTP connection attempts which would be ``len(supported_protocols)`` connections in the worst-case (i.e. if the only thing the server supports is the client's last choice).

**Comment by oberstet** [comment]
Yep, right: RawSocket isn't really negotiation .. the router will just deny a connection. And ype, of course the client can then retry a second time on a new connection with different parameters.

The idea for _not_ having full blown neg. was this:

1. a router should support more (all) variants anyway
2. we don't want to reinvent WebSocket (which is more complex to implement than RawSocket)

Eg WebSocket support payload compression. Now if we start to add feature after feature, the whole point of RawSocket - having a dead simple, minimalistic transport - is lost.

I started RawSocket with the bare, absolute minimum: length prefixed messages. Then a couple of features were added. Adding even more will make it look like WebSocket more and more. And then why do we need it in the first place?
